nixos/gitlab: Fix state directory permissions, clean up initializers directory

talyz opened the pull request:
###### Motivation for this change
Since the preStart script is no longer running in privileged mode, recursively reassign all files in the state directory to the user we're running as. Also, delete the database.yml symlink if it exists,
since we want to create a real file in its place.

I've tested this by manually changing the owner of `gitlab_shell_secret`, and symlinking `database.yml` and `secrets.yml` to an irrelevant file. Additional testing would be appreciated.

The `config/initializers` directory is populated with files from the gitlab distribution on start, but old files will be left in the state folder even if they're removed from the distribution, which can lead to
startup failures. Fix this by always purging the directory on start before populating it.

Should hopefully fix #68696; if so, it needs to be pulled into 19.09 too.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli @globin @aanderse @FRidh

talyz committed:
nixos/gitlab: Fix state directory permissions

Since the preStart script is no longer running in privileged mode, we
reassign the files in the state directory and its config subdirectory
to the user we're running as. This is done by splitting the preStart
script into a privileged and an unprivileged part where the privileged
part does the reassignment.

Also, delete the database.yml symlink if it exists, since we want to
create a real file in its place.

Fixes #68696.

schneefux commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -774,6 +775,10 @@ in {
+            rm '${cfg.statePath}/config/database.yml'
+          fi
+
           ${if cfg.databasePasswordFile != null then ''
```
Not related to this issue, but I noticed that no `database.yml` gets written if the `databasePasswordFile` is not set. So it is impossible to exclusively use `extraDatabaseConfig` to configure the database.
In other modules, it is sometimes easier to ditch the options and put a complete config in the `extraConfig`.
Is that intentional?

talyz replied:
If that's the case, then that's a bug, but I don't have `databasePasswordFile` set and get a `database.yml` anyway (see the else clause of the if expression)

talyz replied:
No problem :)

schneefux commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -774,6 +775,10 @@ in {
 
           ${pkgs.openssl}/bin/openssl rand -hex 32 > ${cfg.statePath}/gitlab_shell_secret
```
I chowned `gitlab_shell_secret` back to `root:root` and it breaks here.
` 4yjg33mxs50ngydn1glpb9zcx15i7np1-unit-script-gitlab-pre-start[24046]: /nix/store/4yjg33mxs50ngydn1glpb9zcx15i7np1-unit-script-gitlab-pre-start: line 14: /var/www/gitlab/gitlab_shell_secret: Permission denied`

talyz replied:
Yes, this hinges on systemd-tmpfiles getting to do its job before the script starts, which it should on reboot or when the configuration is activated, if I'm not mistaken. You can manually try it with `systemd-tmpfiles --create`.

schneefux reviewed (changes requested):
I have tested that this does not break an upgraded installation. If I link the two ymls to random files, they are replaced correctly. 👍 
However, if the shell secret is owned by a different account than `gitlab`, the start script still fails.

schneefux reviewed (approved):
👍

globin commented:
Please don't do this as this takes ages on larger gitlab installations (~15 mins for us)..
I'd prefer fixing this for now with chown and running `preStart` as root again for 19.09 as a compatibility release with the right permissions and switching to `preStart` as gitlab in 20.03

aanderse commented:
If you restored the `tmpfiles` rules from prior to these changes but set `tmpfiles` mode to `z` for the individual files owned by `root` that would save a recursive `chown` + `chmod`, allow you to run the script unprivileged, and not require any further changes at a later time, I believe.

talyz commented:
@globin Oh, for every config activation or just the first one? I guess it makes sense, since it has to stat all files recursively... :/

@aanderse That was my first thought, but it didn't work - systemd-tmpfiles emits the error message
> Detected unsafe path transition /var/gitlab/state → /var/gitlab/state/gitlab_shell_secret during canonicalization of /var/gitlab/state/gitlab_shell_secret.

which it apparently does when a file in a directory owned by an unprivileged user is owned by root...

talyz commented:
Okay, so I've gone with the earlier approach of running a privileged part and an unprivileged part of the `preStart` script, but implemented it locally in the gitlab module rather than the systemd module. Does this work better for you, @globin?

talyz committed:
nixos/gitlab: Clean up the initializers on start

The initializers directory is populated with files from the gitlab
distribution on start, but old files will be left in the state folder
even if they're removed from the distribution, which can lead to
startup failures. Fix this by always purging the directory on start
before populating it.

talyz commented:
I've pushed one additional commit that cleans up the `config/initializers` directory, rebased on master and tested it without any issues. This addresses the concern raised by @bgamari in https://github.com/NixOS/nixpkgs/pull/70216#issuecomment-537646392.

@globin Do you have any issues with the current approach or could this be merged?

aanderse reviewed (approved):
I have not tested, nor ever used the `gitlab` module on NixOS but I understand and approve of the basic idea behind this change. From my limited perspective the actual implementation looks fine as well. :+1:
